app/models/report.rb

```diff
@@ -93,6 +93,208 @@ def cleanup_report
     end
   end
 
+  def self.generate_whole_project_report(project, current_user, current_team)
+    report_contents = []
+    report_contents << {
+      'type_of' => 'project_header',
+      'id' => {
+        'project_id' => project.id
+      },
+      'sort_order' => nil,
+      'children' => []
+    }
+
+    project.experiments.each do |exp|
+      modules = []
+
+      exp.my_modules.each do |my_module|
+        module_children = []
+
+        my_module.protocol.steps.each do |step|
+          step_children = []
+
+          step.assets.each do |asset|
+            step_children << {
+              'type_of' => 'step_asset',
+              'id' => {
+                'asset_id' => asset.id
+              },
+              'sort_order' => nil,
+              'children' => []
+            }
+          end
+          step.tables.each do |table|
+            step_children << {
+              'type_of' => 'step_table',
+              'id' => {
+                'table_id' => table.id
+              },
+              'sort_order' => nil,
+              'children' => []
+            }
+          end
+          step.checklists.each do |step_checklist|
+            step_children << {
+              'type_of' => 'step_checklist',
+              'id' => {
+                'checklist_id' => step_checklist.id
+              },
+              'sort_order' => nil,
+              'children' => []
+            }
+          end
```

Comment:
These three loops could be easily concatenated into two loops. Suppose we defined tuple:
```
tuple = [['assets', 'step_asset'], ['tables, 'step_table'], ['checklists, 'step_checklist']]
tuple.each do |method, name|
  step.send(method).each do |type|
   ...
  end
end
```
Using with the above helper method for hash it would be really concise, not sure how readable though. Probably same tricks apply somewhere below.

Reply:
As mentioned before, time was not my friend for this PR. Will do.